sql/opt/optbuilder: propagate FOR UPDATE locking in scope to ScanExprs

nvb opened the pull request:
Relates to #40205.

This change updates optbuilder to propagate row-level locking modes through the `*scope` object to achieve the same locking clause scoping that Postgres contains. The change propagates the locking mode all the way down to the `ScanExpr` level, where a single locking mode becomes an optional property of a Scan. 

With proper scoping rules, the change also improves upon the locking validation introduced in #43887.

This completes the SELECT FOR [KEY] SHARE/UPDATE optbuilder interactions. The next step will be to hook the ScanExpr locking property into the execbuilder.

nvb committed:
sql/sem/tree: use String methods in Locking{Strength/WaitPolicy} docTable impls

This commit contains a small amount of cleanup to reduce duplication.

Release note: None

nvb committed:
sql/opt/optbuilder: propagate FOR UPDATE locking in scope to ScanExprs

Relates to #40205.

This change updates optbuilder to propagate row-level locking modes on the stack
to achieve the same locking clause scoping that Postgres contains. The change
propagates the locking mode all the way down to the ScanExpr level, where a
single locking mode becomes an optional property of a Scan.

With proper scoping rules, the change also improves upon the locking validation
introduced in #43887.

This completes the SELECT FOR [KEY] SHARE/UPDATE optbuilder interactions. The
next step will be to hook the ScanExpr locking property into the execbuilder.

Release note (sql change): More invalid usages of FOR UPDATE locking clauses
are now rejected by the SQL optimizer.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44015)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 13 of 15 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-:-LyjHOgs8znIVqfr7s-n:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/sql/logictest/testdata/logic_test/select_for_update, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyeZJitARM88qLMhFyS:-LyeZJitARM88qLMhFyT:b-whwscf) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f1295aa4dca48f2f518d721f441fb71d3f4c1b/pkg/sql/logictest/testdata/logic_test/select_for_update#L48)):*
> ```
> 
> # However, we do mirror Postgres in that we require FOR UPDATE targets to be
> # unqualified names and reject anything else.
> ```

Is this needed for correctness? Seems like kind of a random requirement...

Edit -- saw the comment below. Either way is fine with me.
___
*[pkg/sql/logictest/testdata/logic_test/select_for_update, line 180 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyetxtbBzeLXu1hLl0O:-LyetxtbBzeLXu1hLl0P:bxshvvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f1295aa4dca48f2f518d721f441fb71d3f4c1b/pkg/sql/logictest/testdata/logic_test/select_for_update#L180)):*
> ```
> 2
> 
> # Use of SELECT FOR UPDATE requires UPDATE privileges, not just SELECT privileges.
> ```

How about a test of FOR SHARE?
___
*[pkg/sql/opt/optbuilder/scope.go, line 1315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyevrlD3ePidxl9d7K2:-LyevrlEEIaqb446Ar9a:bzf9q2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f1295aa4dca48f2f518d721f441fb71d3f4c1b/pkg/sql/opt/optbuilder/scope.go#L1315)):*
> ```Go
> }
> 
> // lockingForTable returns the desired row-level locking mode for the specifies
> ```

[nit] specifies -> specified
___
*[pkg/sql/opt/optbuilder/scope.go, line 1324 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyewwLW6D67sZXUsbxE:-LyewwLW6D67sZXUsbxF:b78bf2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f1295aa4dca48f2f518d721f441fb71d3f4c1b/pkg/sql/opt/optbuilder/scope.go#L1324)):*
> ```Go
> 	var copied bool
> 	updateRet := func(li *tree.LockingItem) {
> 		if ret == nil && len(li.Targets) == 0 {
> ```

Why do you need the check if `len(li.Targets) == 0` here? Seems like that's already checked below.
___
*[pkg/sql/opt/optbuilder/select.go, line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lyj9sFu1rz1TzfpdZYu:-Lyj9sFu1rz1TzfpdZYv:boudivz) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/select.go#L59)):*
> ```Go
> 
> 		// TODO DURING REVIEW: surprisingly, it looks like all TableName data
> 		// sources are wrapped by an AliasedTableExpr. Is this true? Can we
> ```

As far as I can tell, the parser doesn't directly return `tree.TableName`s, but it does return `tree.TableRef`s (numeric table references).
___
*[pkg/sql/opt/optbuilder/select.go, line 126 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjAMn5EtBMj9nwfe8e:-LyjAMn5EtBMj9nwfe8f:b-608p84) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/select.go#L126)):*
> ```Go
> 		priv := privilege.SELECT
> 		if li != nil {
> 			// SELECT ... FOR [KEY] UPDATE/SHARE requires UPDATE privileges.
> ```

Why does SHARE require update privileges?
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjFBhH8OrT0AfbneIs:-LyjFBhH8OrT0AfbneIt:ber7j3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/testdata/select_for_update#L74)):*
> ```
> 
> build
> SELECT * FROM t FOR UPDATE OF t2
> ```

Is it a problem that we are dropping this locking request? How does Postgres handle this?
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjFiTr69pnGF2zGWgF:-LyjFiTr69pnGF2zGWgG:b-9l1xy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/testdata/select_for_update#L151)):*
> ```
> 
> build
> SELECT * FROM v FOR UPDATE OF v2
> ```

What about `SELECT * FROM v FOR UPDATE OF t`? How should that behave?
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 261 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjGb66E27yd1Kk681K:-LyjGb66E27yd1Kk681L:bc6bkuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/testdata/select_for_update#L261)):*
> ```
>                           ├── columns: a:1(int!null) b:2(int)
>                           └── locking: for-update
> ```

How about adding a test where a stronger locking from outside overrides a weaker locking inside a subquery?
___
*[pkg/sql/sem/tree/select.go, line 1063 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjH8elDeOiUNAuNuVm:-LyjH8elDeOiUNAuNuVn:b2qz76p) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/sem/tree/select.go#L1063)):*
> ```Go
> }
> 
> // Max returns the maximum of the two locking weight policies.
> ```

weight -> wait


<!-- Sent from Reviewable.io -->

nvb committed:
sql/opt/optbuilder: pull WITH processing under buildSelect

This commit restructures the handling of tree.Select and tree.SelectStmt
in the optbuilder. It pulls WITH processing under builer.buildSelect so
that it can merge the ParenSelect traversal used to detect duplicate WITH
clauses with the traversal to detect duplicate ORDER BY and LIMIT clauses.

This will also allow for cleaner plumbing of row-level locking modes in
the next commit.

nvb reviewed (commented):
TFTR Becca! I addressed all of your comments. I also addressed comments that @andy-kimball provided offline.

This led to me pull the locking clauses out of `scope` and passing them on the stack instead. Andy's test case which made it most clear to me that passing locking around on the `scope` was incorrect was the following:
```
SELECT (SELECT a FROM t) FOR UPDATE
```
When we were passing locking information on scope, this would naturally lead to the subquery using row-level locking. I assumed that's what was desired and this case was actually what drove me to use `scope` like this in the first place. It turns out that this isn't desired and that in PG, table `t` is not locked.

So the new approach is simpler and easier to reason about. We just pass the locking specification around to wherever it's needed. I added a commit to cleanup a few of the functions around SelectStmt construction so that we don't need to plumb this locking information all the way through `buildStmt`.

PTAL whenever possible.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-:-LyrpQSV2zQmffR9_T7d:brwi96e)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/select_for_update, line 180 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyetxtbBzeLXu1hLl0O:-Lyrn3J52dniVOLRj5NM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f1295aa4dca48f2f518d721f441fb71d3f4c1b/pkg/sql/logictest/testdata/logic_test/select_for_update#L180)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

How about a test of FOR SHARE?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/scope.go, line 1315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyevrlD3ePidxl9d7K2:-Lyrn7lj5kA0DUnIKqrG:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f1295aa4dca48f2f518d721f441fb71d3f4c1b/pkg/sql/opt/optbuilder/scope.go#L1315)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] specifies -> specified
</blockquote></details>

Removed.
___
*[pkg/sql/opt/optbuilder/scope.go, line 1324 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyewwLW6D67sZXUsbxE:-LyrnBtrDktzdC1rYYVC:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f1295aa4dca48f2f518d721f441fb71d3f4c1b/pkg/sql/opt/optbuilder/scope.go#L1324)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why do you need the check if `len(li.Targets) == 0` here? Seems like that's already checked below.
</blockquote></details>

Removed.
___
*[pkg/sql/opt/optbuilder/select.go, line 59 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lyj9sFu1rz1TzfpdZYu:-LyrnSdK7EksZ_SAmSBK:b6vlgnz) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/select.go#L59)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

As far as I can tell, the parser doesn't directly return `tree.TableName`s, but it does return `tree.TableRef`s (numeric table references).
</blockquote></details>

Huh, I guess that makes sense, but it's certainly not what I was expecting.
___
*[pkg/sql/opt/optbuilder/select.go, line 126 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjAMn5EtBMj9nwfe8e:-Lyrn_OKEvbnJdbqc0AY:b-mpzid4) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/select.go#L126)):*
That's how it works in Postgres for some reason. In https://www.postgresql.org/docs/12/sql-select.html:

> The use of FOR NO KEY UPDATE, FOR UPDATE, FOR SHARE or FOR KEY SHARE requires UPDATE privilege as well

I tested it out as well and that's how it seemed to work.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjFBhH8OrT0AfbneIs:-LyroF6364SO77Brmjg1:b-k7tb9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/testdata/select_for_update#L74)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is it a problem that we are dropping this locking request? How does Postgres handle this?
</blockquote></details>

Postgres throws an error. It's not a huge problem but I do have a pair of TODOs to try to fix it.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjFiTr69pnGF2zGWgF:-LyroMjD0H-IM3qgk2Oz:b-s0rtzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/testdata/select_for_update#L151)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

What about `SELECT * FROM v FOR UPDATE OF t`? How should that behave?
</blockquote></details>

Added a test for that.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 261 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjGb66E27yd1Kk681K:-LyrozF086LzddyLbMxi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/opt/optbuilder/testdata/select_for_update#L261)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

How about adding a test where a stronger locking from outside overrides a weaker locking inside a subquery?
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/select.go, line 1063 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LyjH8elDeOiUNAuNuVm:-Lyrp1zc-txlGWWJ6DIE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ff4c5dabd00a5f8661674d1e4fb1abf44230c7/pkg/sql/sem/tree/select.go#L1063)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

weight -> wait
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44015)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @rytaft)

___
*[pkg/sql/opt/memo/interner.go, line 883 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytNff96AHMmpQ1Vzel:-LytNff96AHMmpQ1Vzem:brjlmfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/18718c4795455caab865aaa306007bac4cb75e8b/pkg/sql/opt/memo/interner.go#L883)):*
> ```Go
> 		return l == r
> 	}
> 	return l.Strength == r.Strength && l.WaitPolicy == r.WaitPolicy
> ```

Better here would be `*l == *r`. If other fields got added to the struct, it would prevent a bug from manifesting if this code wasn't updated.
___
*[pkg/sql/opt/optbuilder/locking.go, line 94 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytZ0eNA0hN0QCtnFHV:-LytZ0eNA0hN0QCtnFHW:b-4b2qzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/18718c4795455caab865aaa306007bac4cb75e8b/pkg/sql/opt/optbuilder/locking.go#L94)):*
> ```Go
> 		return
> 	}
> 	*lm = append(*lm, locking...)
> ```

I think this could mutate the AST slice in case where the existing `lockingSpec` comes from the AST. We want to keep the AST immutable. So instead something like:
```
l := len(*lm)
*lm = append((*lm)[:l:l], locking...)
```
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 10 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytbYIA6Ty4pDp0ypWI:-LytbYIA6Ty4pDp0ypWJ:b-r06wfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L10)):*
> ```
> 
> exec-ddl
> CREATE VIEW v AS SELECT a FROM t
> ```

Should make this a bit trickier by aliasing `t` as `t2` or similar.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 91 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lytbkns31ciiJLw4wVx:-Lytbkns31ciiJLw4wVy:b-6r2bhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L91)):*
> ```
> 
> build
> SELECT * FROM t AS t2 FOR UPDATE OF t
> ```

This test should fail with an error, since `t` is not in-scope.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 123 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lytc_3l8qeQeL7EngBX:-Lytc_3l8qeQeL7EngBY:bw2s2ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L123)):*
> ```
> 
> build
> SELECT * FROM [53 AS t] FOR UPDATE OF t2
> ```

Should be an error case. Perhaps the right way to handle these cases is by doing a FOR UPDATE target scoping check _after_ processing the FROM clause, for sole purpose of triggering errors.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 159 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytcwRT84vWOmwsAdRp:-LytcwRT84vWOmwsAdRq:b3shmm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L159)):*
> ```
> 
> build
> SELECT * FROM v FOR UPDATE OF t
> ```

As above, this should be an error case. I'll stop calling out cases that should be errors from here on out...

I'd also add another case like this, where I try to use `t2`, the alias of `t` in the view:
```
SELECT * FROM v FOR UPDATE of t2
```
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 241 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytdK5XCrAU4N81wGcQ:-LytdK5XCrAU4N81wGcR:b-zes61a) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L241)):*
> ```
> 
> # TODO(nvanbenschoten): To match Postgres perfectly, this would throw an error.
> # It's not clear that it's worth going out of our way to mirror that behavior.
> ```

I think it's worth it because it's confusing to silently make FOR UPDATE a no-op. This syntax looks like it _should_ work, but doesn't, and that's not a great user experience.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 493 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytdqM88IXh_sqeFYM0:-LytdqM88IXh_sqeFYM1:b-i26ccd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L493)):*
> ```
> # Tests with common-table expressions.
> #
> # Unlike with subqueries, row-level locking clauses do not apply to WITH queries
> ```

`Unlike with FROM subqueries`, perhaps?


<!-- Sent from Reviewable.io -->

nvb committed:
sql/opt/optbuilder: validate FOR UPDATE locking targets

This commit addresses a TODO from the previous commit. It adds
logic into the optbuilder to validate that all locking targets
are real relations that are present in the FROM clause of the
SELECT statement.

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44015)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/memo/interner.go, line 883 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytNff96AHMmpQ1Vzel:-LyzkTk07Q8uEBsZNyS5:b7ycmdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/18718c4795455caab865aaa306007bac4cb75e8b/pkg/sql/opt/memo/interner.go#L883)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Better here would be `*l == *r`. If other fields got added to the struct, it would prevent a bug from manifesting if this code wasn't updated.
</blockquote></details>

`tree.LockingItem` isn't comparable because it contains a slice (`Targets`). The slice is guaranteed to be empty (see documentation on ScanPrivate), but it still prevents us from doing `*l == *r`. An alternative approach would be to define a new type and to not use `tree.LockingItem` directly in `ScanPrivate`. I think that's overkill and also comes with its own set of downsides related to leaving room for bugs if anything changes here, but I'd be interested in your take.
___
*[pkg/sql/opt/optbuilder/locking.go, line 94 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytZ0eNA0hN0QCtnFHV:-LyzlaQD8gMkgOK3-dfx:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/18718c4795455caab865aaa306007bac4cb75e8b/pkg/sql/opt/optbuilder/locking.go#L94)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think this could mutate the AST slice in case where the existing `lockingSpec` comes from the AST. We want to keep the AST immutable. So instead something like:
```
l := len(*lm)
*lm = append((*lm)[:l:l], locking...)
```
</blockquote></details>

Good point, done.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 10 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytbYIA6Ty4pDp0ypWI:-LyzmZo_AH0BcGZ9OZRZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L10)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should make this a bit trickier by aliasing `t` as `t2` or similar.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 91 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lytbkns31ciiJLw4wVx:-LyzmciU3cosTRIJNcsh:b-pnhr4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L91)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This test should fail with an error, since `t` is not in-scope.
</blockquote></details>

See below.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 123 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lytc_3l8qeQeL7EngBX:-Lyzm_wlA96bS-sljPb-:b-3bsna4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L123)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should be an error case. Perhaps the right way to handle these cases is by doing a FOR UPDATE target scoping check _after_ processing the FROM clause, for sole purpose of triggering errors.
</blockquote></details>

Discussed on Slack. There was a TODO at the end of `validateLockingInFrom` to address this. We already had access to `fromScope` in `validateLockingInFrom`, but it wasn't immediately clear what the best approach for checking the target scoping would be. I was hoping that `scope` would carry information about the relations it contains, but that isn't the case. `scope` does, however, contain information about the columns it contains. Using this information, I was able to address the TODO in a follow-up commit. PTAL at that approach and let me know if you think there's a better way to accomplish what it's trying to accomplish.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 159 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytcwRT84vWOmwsAdRp:-LyzqREi5zK4KIqljBiy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L159)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

As above, this should be an error case. I'll stop calling out cases that should be errors from here on out...

I'd also add another case like this, where I try to use `t2`, the alias of `t` in the view:
```
SELECT * FROM v FOR UPDATE of t2
```
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 241 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytdK5XCrAU4N81wGcQ:-LyzrIv-BBnSd6qx9Lpx:bhcj06g) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L241)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think it's worth it because it's confusing to silently make FOR UPDATE a no-op. This syntax looks like it _should_ work, but doesn't, and that's not a great user experience.
</blockquote></details>

See the discussion above.
___
*[pkg/sql/opt/optbuilder/testdata/select_for_update, line 493 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytdqM88IXh_sqeFYM0:-LyzrQPSEUhTYzwbsYnI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b0fc83930b5e4f13fb38914ef08b5f6372aed74/pkg/sql/opt/optbuilder/testdata/select_for_update#L493)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

`Unlike with FROM subqueries`, perhaps?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, nice work

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-:-Lz2Uoxj8BcJZdxW1om9:b-qyf0xd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @rytaft)

___
*[pkg/sql/opt/memo/interner.go, line 883 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-LytNff96AHMmpQ1Vzel:-Lz2R_1G-7w5TTbfM0K0:b-lt0lyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/18718c4795455caab865aaa306007bac4cb75e8b/pkg/sql/opt/memo/interner.go#L883)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`tree.LockingItem` isn't comparable because it contains a slice (`Targets`). The slice is guaranteed to be empty (see documentation on ScanPrivate), but it still prevents us from doing `*l == *r`. An alternative approach would be to define a new type and to not use `tree.LockingItem` directly in `ScanPrivate`. I think that's overkill and also comes with its own set of downsides related to leaving room for bugs if anything changes here, but I'd be interested in your take.
</blockquote></details>

I see, makes sense.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r+

rytaft reviewed (commented):
Sorry for the delay, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Just a few typos in comments -- feel free to ignore or address in a future PR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-:-Lz80cuD9H-5EWdWMqDE:b-8rq9cz)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @rytaft)

___
*[pkg/sql/opt/optbuilder/locking.go, line 98 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz8-DtlE6ycn0b9nZhM:-Lz8-DtlE6ycn0b9nZhN:b-3w12bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/locking.go#L98)):*
> ```Go
> }
> 
> // filter returns the desired row-level locking mode for the specifies table as
> ```

specifies -> specified
___
*[pkg/sql/opt/optbuilder/locking.go, line 162 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz8-9quENyuYK5C64tU:-Lz8-9quENyuYK5C64tV:b-5sjxwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/locking.go#L162)):*
> ```Go
> // current scope to the CTE. This mirrors Postgres' behavior. It also avoids a
> // number of awkward questions like how row-level locking would interact with
> // mutating commong table expressions.
> ```

commong -> common
___
*[pkg/sql/opt/optbuilder/select.go, line 138 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz7zv4m8QbOVugs7POE:-Lz7zv4m8QbOVugs7POF:b7qkr87) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/select.go#L138)):*
> ```Go
> 	case *tree.Subquery:
> 		// Remove any target relations from the current scope's locking spec, as
> 		// those only apply to relations in this statements. Interestingly, this
> ```

statements -> statement
___
*[pkg/sql/opt/optbuilder/select.go, line 1232 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz7ykAKAsSYplARtunP:-Lz7ykALAeS9CEtDHh6E:b-lcyesa) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/select.go#L1232)):*
> ```Go
> 			// columns in scope for each of the locking targets. We could use a
> 			// more efficient data structure (e.g. a hash map of relation names)
> 			// to improve the complexity here, but we don't expect the number of
> ```

Did you mean "we expect" instead of "we don't expect"?
___
*[pkg/sql/opt/optbuilder/select.go, line 1262 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz80QGD-ExFzB-DZT_7:-Lz80QGD-ExFzB-DZT_8:b-vln4dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/select.go#L1262)):*
> ```Go
> 
> // raiseLockingContextError raises an error indicating that a row-level locking
> // clause is not permitted in the specified context. locking.set must be true.
> ```

locking.set -> locking.isSet()


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
There wasn't a delay - I should have waited for your stamp as well. I addressed you comments in https://github.com/cockroachdb/cockroach/pull/44169. Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-:-Lz8Y7zi7buTkppRx4sq:b-xmthat)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/optbuilder/locking.go, line 98 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz8-DtlE6ycn0b9nZhM:-Lz8TnNY2xzgi7E845pN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/locking.go#L98)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

specifies -> specified
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/locking.go, line 162 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz8-9quENyuYK5C64tU:-Lz8TpCW1akcMFPjLes4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/locking.go#L162)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

commong -> common
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 138 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz7zv4m8QbOVugs7POE:-Lz8Tt3LECb4GcLm9sJF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/select.go#L138)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

statements -> statement
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 1232 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz7ykAKAsSYplARtunP:-Lz8U20iEUfTSEx-0q0M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/select.go#L1232)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Did you mean "we expect" instead of "we don't expect"?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 1262 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44015#-Lz80QGD-ExFzB-DZT_7:-Lz8UBG_2CKha7IQrKiR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ac7bd34c346688c0d600bb34b7565d810f635c/pkg/sql/opt/optbuilder/select.go#L1262)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

locking.set -> locking.isSet()
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
